src/Microsoft.AspNetCore.SignalR/Internal/AsyncEnumeratorAdapters.cs

```
@@ -0,0 +1,125 @@
+            }
+            else
+            {
+                return (IAsyncEnumerator<object>)enumerator;
+            }
+        }
+
+        private static IAsyncEnumerator<object> BoxEnumerator<T>(IAsyncEnumerator<T> input) where T : struct
+        {
+            return new BoxingEnumerator<T>(input);
+        }
+
+        private class ChannelObserver<T> : IObserver<T>
+        {
+            private WritableChannel<object> _output;
+            private CancellationToken _cancellationToken;
+
+            public ChannelObserver(WritableChannel<object> output, CancellationToken cancellationToken)
+            {
+                _output = output;
+                _cancellationToken = cancellationToken;
+            }
+
+            public void OnCompleted()
+            {
+                _output.TryComplete();
+            }
+
+            public void OnError(Exception error)
+            {
+                _output.TryComplete(error);
+            }
+
+            public void OnNext(T value)
+            {
+                _cancellationToken.ThrowIfCancellationRequested();
+
+                // This will block the thread emitting the object if the channel is bounded and full
+                // I think this is OK, since we want to push the backpressure up. However, we may need
+                // to find a way to force the entire subscription off to a dedicated thread in order to
+                // ensure we don't block other tasks
+
+                // Right now however, we use unbounded channels, so all of the above is moot because TryWrite will always succeed
+                while (!_output.TryWrite(value))
+                {
+                    // Wait for a spot
+                    if (!_output.WaitToWriteAsync(_cancellationToken).Result)
+                    {
+                        // Channel was closed.
+                        throw new InvalidOperationException("Output channel was closed");
```

It will crash the thing that drives the Observer, which depends on the underlying implementation of IObservable. We could suppress the exception and just drop items I guess.